AD Function API

dschwen opened the pull request:
## Reason
We need to be able to evaluate Moose Functions in AD enabled classes. 

## Design
Add a _overload_ of `Function::value` that takes and returns dual numbers. For maximum compatibility this is implemented in the base class using the chain-rule method, where we set the `derivatives()` using `timeDerivative` and `gradient` data.

Derived Function classes are free to override the AD overload. In this PR I did that for the _piecewise multi_ Functions.

This PR also cleans up the `PiecewiseMultiInterpolation` class hierarchy and introduces a semidynamic vector class that avoids excessive heap allocation/deallocation. I can be used for small vectors with a known `max_size()`. Here it is used for a 1 - 4 dimensional index/point.

## Impact
Added capability.

Closes #20193

dschwen committed:
Add semi-dynamic vector class (#20193)

dschwen committed:
Use semi dynamic vectors (#20193)

dschwen committed:
Add AD Function API (#20193)

dschwen committed:
Utilize AD Function API (#20193)

dschwen commented on modules/heat_conduction/src/materials/ElectricalConductivity.C:
```diff
@@ -43,55 +44,25 @@ ElectricalConductivityTempl<is_ad>::ElectricalConductivityTempl(const InputParam
-    temp_qp = _T[_qp];
-  }
+  const auto & temp_qp = _T[_qp];
 
-  const Real resistivity = _ref_resis * (1.0 + _temp_coeff * (temp_qp - _ref_temp));
+  const auto resistivity = _ref_resis * (1.0 + _temp_coeff * (temp_qp - _ref_temp));
   const Real dresistivity_dT = _ref_resis * _temp_coeff;
   _electric_conductivity[_qp] = 1.0 / resistivity;
-  _delectric_conductivity_dT[_qp] = -1.0 / (resistivity * resistivity) * dresistivity_dT;
-
-  if (is_ad)
-    setDerivatives(_electric_conductivity[_qp], _delectric_conductivity_dT[_qp], _ad_T[_qp]);
+  _delectric_conductivity_dT[_qp] =
+      -1.0 / Utility::pow<2>(MetaPhysicL::raw_value(resistivity)) * dresistivity_dT;
```
Here is a good example how the new AD `Function::value` API simplifies end user code

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/946538/) on 67594f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20200/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20200/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 752b49c85cdc75523552b245c55005fd88010d9a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/946538/) on 67594f2 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen commented on modules/thermal_hydraulics/src/userobjects/ADShaftConnectedCompressor1PhaseUserObject.C:
```diff
@@ -185,12 +185,12 @@ ADShaftConnectedCompressor1PhaseUserObject::computeFluxesAndResiduals(const unsi
-    const Real y2_Rp = _Rp_functions[x2]->value(MetaPhysicL::raw_value(flow_rel_corr), Point());
+    const auto y1_Rp = _Rp_functions[x1]->value(flow_rel_corr, ADPoint());
+    const auto y2_Rp = _Rp_functions[x2]->value(flow_rel_corr, ADPoint());
     const ADReal Rp_m = (y2_Rp - y1_Rp) / (x2_spd - x1_spd);
     const ADReal Rp = y1_Rp + (speed_rel_corr - x1_spd) * Rp_m;
-    const ADReal y1_eff = _eff_functions[x1]->value(MetaPhysicL::raw_value(flow_rel_corr), Point());
-    const ADReal y2_eff = _eff_functions[x2]->value(MetaPhysicL::raw_value(flow_rel_corr), Point());
+    const ADReal y1_eff = _eff_functions[x1]->value(flow_rel_corr, ADPoint());
+    const ADReal y2_eff = _eff_functions[x2]->value(flow_rel_corr, ADPoint());
```
here we are now getting derivatives that were dropped before

joshuahansel replied:
Yep, addressing this was one of my commits in https://github.com/idaholab/moose/pull/20217. Glad to see there's a better way now!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/946606/) on 4e9eb1f wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen committed:
Fix compile issues

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/946708/) on b7c0bc9 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/946924/) on cade74c wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/946941/) on cade74c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20200/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947228/) on c79ae90 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen commented:
That flipping "overloaded virtual" warning pops up on civet, but not on my local machine. I know the fix is adding a `using Function::value;`, but that's really not super elegant.

dschwen committed:
Each interpolation class should support both AD and non-AD (#20193)

lindsayad commented on framework/include/functions/Function.h:
```diff
@@ -64,6 +65,16 @@ class FunctionTempl : public MooseFunctionBase,
+   * \param p The Point in space (x,y,z)
+   * \return A scalar of the function evaluated at the time and location
+   */
+  virtual ADReal value(ADReal t, const ADPoint & p) const;
```
```suggestion
  virtual ADReal value(const ADReal & t, const ADPoint & p) const;
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947443/) on a8e7273 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

lindsayad commented on framework/include/utils/MooseUtils.h:
```diff
@@ -983,6 +984,48 @@ buildRequiredMaterials(const Consumers & mat_consumers,
+ * and a chosen dynamic size. This container avoids heap allocation and
+ * is meant as a replacement for small local std::vector variables.
+ */
+template <typename T, std::size_t N>
+class SemidynamicVector
```
This looks almost exactly like `MetaPhysicL::DynamicStdArrayWrapper` with some slight improvements in a couple of places, e.g. I prefer your implementation of `end()` because it removes an operation. Perhaps we could improve and reuse the current class? I know that this will make it slower to get it available in MOOSE...

dschwen replied:
Yeah, good idea. Swapping the use of this class out would involve changing 3 typedefs. I can check if `MetaPhysicL::DynamicStdArrayWrapper` works out of the box. If so I'll scrap my implementation and a PR to change end() could be made at any point in time.

lindsayad replied:
You could alias `DynamicStdArrayWrapper` to a name you like better

dschwen replied:
is there maybe some advanced reason for using `_data.end() - (N - _dynamic_n);` rather than `_data.begin() + _dynamic_n;` ?

lindsayad replied:
I do not think so, but I suppose that I cannot say for sure until we try it out. It's possible that past Alex did this for a good reason, but current Alex thinks he probably just overlooked this.

lindsayad replied:
Did it not work out of the box? I feel as if you're getting pretty far down the road here...

dschwen replied:
Sorry, I haven't even touched that part again yet, there was so much other stuff on this PR that kept me busy :-). Any comments regarding my last comment above?

dschwen replied:
By the way, is the `NType` a clever way of avoiding non-type template parameters?

dschwen replied:
`DynamicStdArrayWrapper::DynamicStdArrayWrapper(std::size_t n)` does not exist. And The MooseADWrapper template that works for `std::array` doesn't apply (because of the `NType`  vs. `std::size_t` template argument). All that could be worked around. But color me concerned about the `MooseADWrapper<DynamicStdArrayWrapper>::type` issue.

dschwen replied:
I suppose I could change my class to derive from DynamicStdArrayWrapper and implement that ctor like

```
template <std::size_t N>
class SemidynamicVector : public DynamicStdArrayWrapper<NWrapper<N>>
{
  SemidynamicVector(std::size_t n) : _dynamic_n(n) {}
  // possibly some 'using' stuff
};
```

dschwen replied:
Yeah, good point. `std::vector` uses _zero initialization_, while _std::array_ uses _default initialization_ (meaning no initialization for built-ins). If the class is supposed to be a drop in for _std::vector_, we should add an `assign` call to the ctor.

lindsayad replied:
`ADReal` is a `DualNumber<Real, SemiDynamicSparseNumberArray<Real, dof_id_type, NWrapper<n>>, true>`, in other words an instantiation of a `template <typename, typename, bool>` so there are no dangers of your template specialization matching `DualNumber` itself

dschwen replied:
sure, but I could still at least derive for your container. It is a bit more comprehensive than my implementation anyways with reverse iterators and move construction.

lindsayad replied:
At this point I would say up to you. I'll accept either choice

dschwen replied:
Ok, I have a derived version ready, but `_data` is private, so I cannot add my initialization. I'll just leave it at _default initialization_ for now and put in a PR to make the members protected, if that's fine with you.

lindsayad replied:
sounds good to me!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947690/) on 2753294 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen committed:
Fix more compile issues (#20193)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947853/) on d339749 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947890/) on ff3af57 wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen committed:
Update VPP function (#20193)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947924/) on a06f8eb wanted to post the following:

Warning: This PR changes repo size by 5.43 MiB.

dschwen committed:
Fix more compile issues (#20193)

dschwen committed:
Backwards compatibility fix (#20193)

dschwen committed:
Fix jacobian (#20193)

dschwen committed:
Avoid division by zero in thermal hydraulics (#20193)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/949043/) on 8623eeb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.29%</td>
      <td>82.26%</td>
      <td>-0.03%</td>
      <td>61.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73471</td>
      <td>73482</td>
      <td>+11</td>
      <td>92</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15809</td>
      <td>15849</td>
      <td>+40</td>
      <td>58</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/framework)


## Modules coverage

### Functional expansion tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.27%</td>
      <td>83.12%</td>
      <td>-0.15%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>926</td>
      <td>926</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>186</td>
      <td>188</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/functional_expansion_tools.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/functional_expansion_tools)

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.77%</td>
      <td>89.56%</td>
      <td>+0.79%</td>
      <td>67.74%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2997</td>
      <td>2994</td>
      <td>-3</td>
      <td>21</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>379</td>
      <td>349</td>
      <td>-30</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/heat_conduction.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/heat_conduction)

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.02%</td>
      <td>87.29%</td>
      <td>-0.73%</td>
      <td>50.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>316</td>
      <td>316</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>43</td>
      <td>46</td>
      <td>+3</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/level_set.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/level_set)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.23%</td>
      <td>90.20%</td>
      <td>-0.03%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11978</td>
      <td>11978</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1297</td>
      <td>1301</td>
      <td>+4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/phase_field.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/phase_field)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.42%</td>
      <td>89.42%</td>
      <td>+0.00%</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26139</td>
      <td>26131</td>
      <td>-8</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3093</td>
      <td>3092</td>
      <td>-1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/tensor_mechanics)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c98ae8</th>
      <th colspan=4>#20200 8623ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.34%</td>
      <td>90.34%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14918</td>
      <td>14918</td>
      <td>-</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1596</td>
      <td>1596</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage_diff/thermal_hydraulics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20200/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20200/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 61.33% is less than the suggested 90.0%
- `functional_expansion_tools` new line coverage rate 0.00% is less than the suggested 90.0%
- `heat_conduction` new line coverage rate 67.74% is less than the suggested 90.0%
- `level_set` new line coverage rate 50.00% is less than the suggested 90.0%
- `phase_field` new line coverage rate 0.00% is less than the suggested 90.0%
- `tensor_mechanics` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/949043/) on 8623eeb wanted to post the following:


The following coverage requirement(s) failed:
- `level_set` coverage rate 87.29% is less than the required 88.0%

dschwen committed:
Update bicubic interpolation (#20193)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/949087/) on 61e84df wanted to post the following:


The following coverage requirement(s) failed:
- `level_set` coverage rate 87.29% is less than the required 88.0%

dschwen committed:
Regold thermal hydraulics test (#20193)

lindsayad commented on framework/include/utils/MooseADWrapper.h:
```diff
@@ -25,16 +25,24 @@ struct MooseADWrapperStruct<Real, true>
   typedef W<typename MooseADWrapperStruct<T, is_ad>::type> type;
 };
 
-template <typename T, bool is_ad>
-struct MooseADWrapperStruct<std::vector<T>, is_ad>
```
I guess `std::vector` isn't actually currently being used anywhere? Because this new struct surely won't match

dschwen replied:
Doesn't https://github.com/idaholab/moose/pull/20200/files#diff-97b9b30d08fb2ef8e6d28995cd9a02639e2f0f1aa94cb0d8e77380d3d1092284R34-R39 already match `std::vector<T>`?!

dschwen replied:
`DynamicStdArrayWrapper::DynamicStdArrayWrapper(std::size_t n)` does not exist. And The MooseADWrapper template that works for `std::array` doesn't apply (because of the `NType`  vs. `std::size_t` template argument). All that could be worked around. But color me concerned about the `MooseADWrapper<DynamicStdArrayWrapper>::type` issue.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/949478/) on d70fa4d wanted to post the following:


The following coverage requirement(s) failed:
- `level_set` coverage rate 87.29% is less than the required 88.0%

dschwen committed:
Generic qpoint and AD test (#20193)

lindsayad commented on framework/include/utils/MooseADWrapper.h:
```diff
@@ -25,16 +25,24 @@ struct MooseADWrapperStruct<Real, true>
+  typedef ADPoint type;
+};
+
+// W<T> -> W<ADT>, e.g. std::vector<Real> -> std::vector<ADReal>
 template <template <typename> class W, typename T, bool is_ad>
```
Yea I don't think this would match `std::vector` because `std::vector` is a `template <typename, typename>`

dschwen committed:
Use metaphysicl class (#20193)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/950949/) on cc8edb6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20200/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20200/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c068d4e54568d8e83799c0c3de2328ae33c09dd4`

lindsayad committed:
Use Newton solve type for level set test

Unlike global AD indexing, local AD indexing respects a user's coupling
matrix. The default for PJFNK is block-diagonal whereas the default for
NEWTON is full

lindsayad reviewed (changes requested):
Could we get a test of the new base class `Function::value` overload? That looks like the primary reason for the "low" coverage, and that's a pretty important function to cover I think as it's at the crux of the PR. Other than that, I think this looks great

dschwen commented on framework/include/utils/MooseUtils.h:
```diff
@@ -986,6 +989,43 @@ buildRequiredMaterials(const Consumers & mat_consumers,
+ * and a chosen dynamic size. This container avoids heap allocation and
+ * is meant as a replacement for small local std::vector variables.
+ * Note: this class uses default initialization, which will not initialize built-in types.
+ * Note: to to an assertion bug in DynamicStdArrayWrapper we have to allocate one element more
```
Yeah, still needed

dschwen committed:
Add coupled value function material (#20193)

dschwen committed:
Add ad function framework test (#20193)

recuero commented on framework/doc/content/source/materials/CoupledValueFunctionMaterial.md:
```diff
@@ -0,0 +1,14 @@
+# CoupledValueFunctionMaterial
+
+!syntax description /Materials/CoupledValueFunctionMaterial
+
```
Does the capability introduced in this PR has to be necessarily used from a material?

dschwen replied:
No, I modified ADBodyForce and added tests for it. That shows how to use it from a kernel.

recuero commented on framework/doc/content/source/materials/CoupledValueFunctionMaterial.md:
```diff
@@ -0,0 +1,14 @@
+# CoupledValueFunctionMaterial
+
+!syntax description /Materials/CoupledValueFunctionMaterial
+
```
Does the capability introduced in this PR have to be necessarily used from a material?

recuero commented on framework/include/utils/MooseUtils.h:
```diff
@@ -986,6 +989,43 @@ buildRequiredMaterials(const Consumers & mat_consumers,
+  {
+    // const auto old_dynamic_n = Parent::size();
+    Parent::resize(new_size);
+    // TODO: uncomment this once https://github.com/libMesh/MetaPhysicL/pull/8
+    //       makes it all the way into MOOSE.
+    // for (const auto i : make_range(old_dynamic_n, _dynamic_n))
+    //   _data[i] = {};
+  }
```
Will it take long? Just wondering if we'll forget.

dschwen replied:
It's already merged into metaphysicl. Next libmesh update should add it. Probably a two/three weeks.

recuero commented on framework/src/materials/CoupledValueFunctionMaterial.C:
```diff
@@ -0,0 +1,90 @@
+                                  "If provided, an entry per couple variable specifies "
+                                  "the function argument it should apply to.");
+  params.addRequiredParam<MaterialPropertyName>("prop_name", "Output property name");
+  params.addCoupledVar(
+      "v",
+      "List of up to four coupled variables that are substituted for x,y,z, and t "
+      "in the coupled function (or the order chosen using the `parameter_order` parameter)");
```
Can I devise a function of four parameters that have nothing to do with x,y,z, and t? (I assume so).

dschwen replied:
Yeah, a parsed function for example, like I did in the test. Xyzt are just placeholders.

recuero commented on framework/src/materials/CoupledValueFunctionMaterial.C:
```diff
@@ -0,0 +1,90 @@
+    const InputParameters & parameters)
+  : Material(parameters),
+    _prop(declareGenericProperty<Real, is_ad>(getParam<MaterialPropertyName>("prop_name"))),
+    _function(getFunction("function")),
+    _vals(coupledGenericValues<is_ad>("v")),
+    _nvals(coupledComponents("v"))
```
Where do these variables need to live? Would this work for boundary restricted variables?

dschwen replied:
This is not different from any other material. So I suppose yes.

recuero commented on test/tests/materials/coupled_value_function/adjac.i:
```diff
@@ -0,0 +1,91 @@
+[]
+
+[Materials]
+  [Du]
+    type = ADCoupledValueFunctionMaterial
+    function = x
+    v = v
+    prop_name = Du
+  []
```
I don't think we can pull a material like this from a mortar constraint object.

recuero replied:
Nothing. Just mentioning that per our conversation on using functions directly, instead of materials, to include derivative information with multivariable interpolation.

dschwen replied:
Oh, yeah, you can just couple a function into a constraint and evaluate it with coupled variables.

joshuahansel reviewed (changes requested):
Can you add documentation to the Functions system page and then a newsletter entry?

dschwen committed:
Add simplified function interface (#20193)

dschwen committed:
Add docs and newsletter (#20193)

dschwen commented:
> Can you add documentation to the Functions system page and then a newsletter entry?

Done

recuero commented on modules/doc/content/newsletter/2022_02.md:
````diff
@@ -74,6 +74,42 @@ to full compatibility with the current kernels.  Some features the Lagrangian sy
+```
+
+This simplifies code that uses function evaluations with custom variables, such
+as in `HeatConductionMaterial`, where a function can be supplied which only
+depends on time, and Temperature is supplied as the time argument.
````
Don't fully understand. Does the function only depend on time or does temperature replace the time argument?

dschwen replied:
what's called "time" internally in the function system is not necessarily the actual simulation time, but whatever the caller substituted in for the first argument. Yeah, I get it, it it confusing. If you have a suggestion on how to word it more clearly I'd be happy to adopt it.

recuero replied:
"where a function that only depends on temperature can be supplied. The variable temperature takes the place of the first argument (internally designated as ``time'')."?

If I understood that.

dschwen commented:
hmm, _with_ teh PR the points selected in ``

are
```
(x,y,z)=(       0,        0,        0)
(x,y,z)=(       0,        0,        0)
(x,y,z)=(       0,        0,        0)
(x,y,z)=(       0,        0,        0)
```

without the PR they are

```
(x,y,z)=(    0.25,     0.25,        0)
(x,y,z)=(    0.25,    -0.25,        0)
(x,y,z)=(   -0.25,     0.25,        0)
(x,y,z)=(   -0.25,    -0.25,        0)
```

I may have introduced a subtle bug in the `PiecewiseLinear` Function.

dschwen commented:
LOL, I got it. The Mastodon dirackernel calls the function like this:

```
      _p(0) = function_x->value(i + 1, _qp);
```

this was probably written thinking that `_qp` is the quadrature point. However it is only the index, and it got happily casted to a `Point`, but now it instantiates a `value(U t, U x)` call with `U` being an integer. The sampled value of 0.25 then gets rounded to 0.

dschwen committed:
Permit use of integer arguments (#20193)

dschwen commented on framework/doc/content/source/materials/CoupledValueFunctionMaterial.md:
```diff
@@ -0,0 +1,14 @@
+
+CoupledValueFunctionMaterial evaluates MOOSE Functions with a set of up to
+four coupled variable values `v` as its input parameters. The coupled variable
+values are substituted for the `x`,`y`,`z`, and `t` function variables in that
```
Hm. The xyzt order has been used in various MOOSE objects. It allows the most compact assignment 

```
if (i < 3)
  p(i) = value
else
 t = value
```

but you are right, it contrasts the txyz order I added for the new API. That order was motivated by the fact that we have lots of evaluations that only use the time argument, and also the legacy time (t), location (x,y,z) order.

Let me know if you'd rather have the default order changed, either way is fine with me.

recuero replied:
If you are going to push again, I'd add to the doc the ability to select the variable order from the input file.

dschwen replied:
Yeah, I will. Just waiting to see if the auto return type passes our minimum compiler requirements.

dschwen committed:
Betetr handling of degenerate interpolations (#20193)

dschwen commented:
![image](https://user-images.githubusercontent.com/202302/153656504-3949453e-f1e6-4837-bff0-f0e0a4d8035a.png)

Huh?! I _just_ rebased on idaholab/next! The last commit I see touching Assembly.C was 8/21. What's going here?

lindsayad commented:
we merged #18563 last night which definitely modified `Assembly.C`

dschwen commented:
> we merged #18563 last night which definitely modified `Assembly.C`

Yeah, I didn't see that in my history after rebasing though. 

Anyhow, after pushing again it's fine. Pronghorn fail is unrelated, right?
